compiler/incremental-compilation-impl/src/org/jetbrains/kotlin/incremental/classpathDiff/ClasspathSnapshotter.kt

```diff
@@ -108,114 +179,57 @@ object ClassSnapshotter {
+    /** The directory or jar to read. */
+    private val root: Path
 
-            val inaccessible = if (isInaccessible()) {
-                true
-            } else classId.outerClassId?.let { outerClassId ->
-                classIdToClassInfo[outerClassId]?.isTransitivelyInaccessible()
-                // If we can't find the outer class from the given list of classes (which could happen with faulty jars), we assume that
-                // the class is accessible (see this method's kdoc).
-                    ?: false
-            } ?: false
+    private val closer = Closer.create()
 
-            return inaccessible.also {
-                classIsInaccessible[this] = inaccessible
-            }
+    init {
+        root = if (directoryOrJar.isDirectory) {
+            directoryOrJar.toPath()
+        } else {
+            check(directoryOrJar.isFile && directoryOrJar.path.endsWith(".jar", ignoreCase = true))
+            FileSystems.newFileSystem(directoryOrJar.toPath(), null as ClassLoader?)
+                .also { closer.register(it) }
+                .rootDirectories.single()
         }
-
-        return classesInfo.filterTo(mutableSetOf()) { it.isTransitivelyInaccessible() }
     }
-}
-
-/** Utility to read the contents of a directory or jar. */
-private object DirectoryOrJarContentsReader {
 
     /**
-     * Returns a map from Unix-style relative paths of entries to their contents. The paths are relative to the given container (directory
-     * or jar).
-     *
-     * The map entries need to satisfy the given filter.
+     * Returns the Unix-style relative paths of all entries under the containing directory or jar which satisfy the given [filter].
      *
-     * The map entries are sorted based on their Unix-style relative paths (to ensure deterministic results across filesystems).
+     * The paths are in Unix style and are sorted to ensure deterministic results across platforms.
      *
-     * Note: If a jar has duplicate entries after filtering, only the first one is retained.
+     * If a jar has duplicate entries, only unique relative paths are kept in the returned list.
      */
-    fun read(
-        directoryOrJar: File,
-        entryFilter: ((unixStyleRelativePath: String, isDirectory: Boolean) -> Boolean)? = null
-    ): LinkedHashMap<String, ByteArray> {
-        return if (directoryOrJar.isDirectory) {
-            readDirectory(directoryOrJar, entryFilter)
-        } else {
-            check(directoryOrJar.isFile && directoryOrJar.path.endsWith(".jar", ignoreCase = true))
-            readJar(directoryOrJar, entryFilter)
-        }
+    fun getUnixStyleRelativePaths(filter: (unixStyleRelativePath: String, isDirectory: Boolean) -> Boolean): List<String> {
+        return Files.walk(root).use {
+            // Use `Set` to de-duplicate entries
+            it.collect(Collectors.toSet())
+        }.map {
+            val relativePath = it.root.relativize(it).toString()
```

Comment:
Can we use https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.io.path/java.nio.file.-path/invariant-separators-path-string.html?

Reply:
Nice -- I tried that but got "unresolved reference" (looks like we don't have access to `kotlin.io.path` from here).